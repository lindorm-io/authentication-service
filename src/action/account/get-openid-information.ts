import { IAuthContext } from "../../typing";
import { Scope, TokenIssuer, isScope } from "@lindorm-io/jwt";
import { InvalidScopeError } from "../../error";

export interface IGetOpenIdInformation {
  email: string;
  emailVerified: boolean;
  sub: string;
  updatedAt: number;
}

export const getOpenIdInformation = (ctx: IAuthContext) => async (): Promise<IGetOpenIdInformation> => {
  const { account, logger, token } = ctx;
  const {
    bearer: { scope },
  } = token;

  logger.info("requesting openid account information", { id: account.id });

  if (!isScope(scope, Scope.DEFAULT)) {
    throw new InvalidScopeError(scope, Scope.DEFAULT);
  }

  if (!isScope(scope, Scope.OPENID)) {
    throw new InvalidScopeError(scope, Scope.OPENID);
  }

  return {
    email: account.email,
    emailVerified: true,
    sub: account.id,
    updatedAt: TokenIssuer.dateToExpiry(account.updated),
  };
};
