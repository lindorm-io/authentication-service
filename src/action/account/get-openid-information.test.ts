import MockDate from "mockdate";
import { getOpenIdInformation } from "./get-openid-information";
import { getTestAccount, logger } from "../../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../support", () => ({
  assertBearerTokenScope: jest.fn(() => () => undefined),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("getOpenIdInformation", () => {
  let ctx: any;

  beforeEach(() => {
    ctx = {
      account: getTestAccount("email@lindorm.io"),
      logger,
    };
  });

  test("should return openid information", async () => {
    await expect(getOpenIdInformation(ctx)()).resolves.toMatchSnapshot();
  });
});
