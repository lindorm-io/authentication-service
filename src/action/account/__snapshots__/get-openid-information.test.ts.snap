// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOpenIdInformation should return openid information 1`] = `
Object {
  "email": "email@lindorm.io",
  "emailVerified": true,
  "sub": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "updatedAt": 1577862000,
}
`;
